### ISSUE TITLE
`sqlalchemy.sql.elements.BinaryExpression` `compare` reports False when expression involves a coalesce

### ISSUE THREAD

**Issue by knguyen1** [sqlalchemy/sqlalchemy#8590]
### Describe the bug

Trying to `unittest` some `sqlalchemy` filter expressions and currently failing.  You can reproduce below.  My expressions `bin3`/`bin4` are identical but `compare` is reporting `False`.

### To Reproduce

```python
>>> import sqlalchemy as sa
>>> class OrmFoo:
...     id = sa.Column(sa.Integer)
...     status = sa.Column(sa.String(10))
...
>>> bin1 = OrmFoo.id == 1
>>> bin2 = OrmFoo.id == 1
>>> bin1.compare(bin2)
True
>>> from sqlalchemy.sql.functions import coalesce
>>> bin3 = coalesce(OrmFoo.status, "foo") != "bar"
>>> bin4 = coalesce(OrmFoo.status, "foo") != "bar"
>>> bin3.compare(bin4)
False
>>> bin3
<sqlalchemy.sql.elements.BinaryExpression object at 0x7ffb6ef2bc10>
>>> bin4
<sqlalchemy.sql.elements.BinaryExpression object at 0x7ffb6ef2be10>
>>> bin1
<sqlalchemy.sql.elements.BinaryExpression object at 0x7ffb71edf150>
>>> bin2
<sqlalchemy.sql.elements.BinaryExpression object at 0x7ffb6ef2ba10>
>>> sa.__version__
'1.3.24'
```

### Error

```
# Copy complete stack trace and error message here, including SQL log output if applicable.
``` 
N/A

### Versions

- OS: Amazon Linux 2
- Python: Python 3.7.10
- SQLAlchemy: 1.3.24
- Database: N/A
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): N/A

**Comment by knguyen1** [comment]
Not ideal.  But I was able to assert the `str` representations.  e.g.

`assert str(bin3) == str(bin4)`

**Comment by zzzeek** [comment]
hey there 

.compare() was an internal only method in 1.3 and earlier and was never fully implemented until #4336, which was released in version 1.4, where it is now central to how Core works and the example you have above has no issues.   so you will want to upgrade.
